README.md
# Color Hash

## A utility for generating persistent colors based on a string for flutter.

![logo](https://github.com/alekangelov/color_hash/raw/main/logo.jpg)

### Motivation

I needed a way to generate a color based on a string that would be consistent and persistent. Flutter has it's own comparison method so I opted to use that. There are all the features you would want from a color generator, including the ability to generate create a factory with your own options.

### Usage

You can create a factory or use it with the default options

```dart
import 'package:color_hash/color_hash.dart';

void defaultOptions() {
  ColorHash colorHash = ColorHash("input");
  Color color = colorHash.toColor();
  print(color); // Color(0xFFSomeColor)
}

void hasherFactory() {
  final hasher = ColorHash.createHasher(
    saturation: 0.5,
    lightness: 0.5,
    hue: (0, 360),
  );

  final hashed = hasher("input");
  print(hashed.toColor()); // Color(0xFFSomeColor)
}

```

### Tooling

I also made this app that lets you dial in the settings and see the output:
![screenshot of tool](https://github.com/alekangelov/color_hash/raw/main/screenshot.jpg)
https://alekangelov.github.io/color_hash/

### Options

| key        | type             | description                                                                                      |
| ---------- | ---------------- | ------------------------------------------------------------------------------------------------ |
| 0          | String           | The original input is a positional argument and is used to create the color for the whole thing. |
| saturation | double           | How saturated should the output color be.                                                        |
| lightness  | double           | How dark/light should the output color be.                                                       |
| hue        | (double, double) | What hue range should the output color be (between 0 and 360, otherwise it will throw).          |

### Author

Alek Angelov - [GitHub](https://github.com/alekangelov) - [LinkedIn](https://www.linkedin.com/in/alekangelov/) - [Website](https://alekangelov.com)

### Licence - MIT

Copyright 2023 Alek Angelov

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## 1.0.1

- Change readme to use absolute links for pub.dev

## 1.0.0

- ColorHash function generates a color based on a hash (string)


test/color_hash_test.dart
import 'dart:ui';

import 'package:color_hash/color_hash.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const baseString = "Some string";
  test('same string hashes should yield same colors', () {
    final hashed = ColorHash(baseString);
    final secondHashed = ColorHash("Some string");
    final thirdHashed = ColorHash("Some other string");
    expect(hashed.toColor(), secondHashed.toColor());
    expect(hashed.toColor(), isNot(thirdHashed.toColor()));
  });

  test('should give a hex string if needed', () {
    final hashed = ColorHash(baseString);
    expect(hashed.toHexString(), "#4055BF");
  });

  test('create hasher should work properly', () {
    final hasher = ColorHash.createHasher(
      lightness: 0.9,
      saturation: 0.25,
      hue: (20, 30),
    );

    expect(hasher(baseString), const Color(0xffece3df));
  });

  test('should throw if invalid values are given', () {
    expect(
      () => ColorHash(baseString,
          hue: (555, 555), saturation: 1.1, lightness: -0.1),
      throwsAssertionError,
    );
    expect(
        () => ColorHash.createHasher(
              lightness: 2,
              saturation: 3,
              hue: (20, 578),
            ),
        throwsArgumentError);
  });
}


example/lib/main.dart
import 'package:color_hash/color_hash.dart';
import 'package:flutter/material.dart';
import 'package:url_launcher/link.dart';

void main() {
  // just call the clas with a string input and it will
  // produce a color

  final hashed = ColorHash(
    "Some string", // initial hash (user id or something)
    lightness: 0.5, // lightness of the color
    saturation: 0.5, // saturation of the color
    hue: (0, 360), // hue range
  );

  hashed.toColor(); // Color(0xff4055bf)

  runApp(const App());
}

// this here is the demo website where you can
// generate your own properties and see the result

class MaxWidth extends StatelessWidget {
  const MaxWidth({required this.child, super.key});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ConstrainedBox(
        constraints: const BoxConstraints(
          maxWidth: 512,
        ),
        child: child,
      ),
    );
  }
}

class App extends StatefulWidget {
  const App({super.key});

  @override
  State<App> createState() => _AppState();
}

class _AppState extends State<App> {
  final TextEditingController _controller = TextEditingController(text: "Seed");

  @override
  void initState() {
    super.initState();
    _controller.addListener(() {
      setState(() {
        _hash = _controller.text;
      });
    });
  }

  String _hash = "Seed";
  (double, double) _hue = (0, 360);
  double _saturation = 0.5;
  double _lightness = 0.6;
  bool? _dark;

  @override
  Widget build(BuildContext context) {
    const spacing = SliverPadding(padding: EdgeInsets.all(20));
    final hashed = ColorHash(
      _hash,
      lightness: _lightness,
      saturation: _saturation,
      hue: _hue,
    );
    final hsv = HSVColor.fromColor(hashed.toColor());
    final currentBrightness = _dark == null
        ? MediaQuery.of(context).platformBrightness
        : _dark!
            ? Brightness.dark
            : Brightness.light;
    return MaterialApp(
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(
          seedColor: hashed.toColor(),
          brightness: currentBrightness,
        ),
        useMaterial3: true,
      ),
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: CustomScrollView(
          slivers: [
            SliverAppBar(
              title: const Text(
                'Color Hash',
              ),
              centerTitle: true,
              actions: [
                IconButton(
                  onPressed: () {
                    setState(() {
                      _dark = !(currentBrightness == Brightness.dark);
                    });
                  },
                  icon: Icon(
                    currentBrightness == Brightness.dark
                        ? Icons.dark_mode
                        : Icons.light_mode,
                  ),
                )
              ],
            ),
            SliverToBoxAdapter(
              child: MaxWidth(
                child: Padding(
                  padding: const EdgeInsets.all(20),
                  child: TextField(
                    controller: _controller,
                    decoration: const InputDecoration(
                      border: OutlineInputBorder(),
                      labelText: 'Seed string',
                    ),
                  ),
                ),
              ),
            ),
            spacing,
            SliverToBoxAdapter(
              child: MaxWidth(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.start,
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 20),
                      child: Text(
                        "Hue range: ${_hue.$1.toInt()} - ${_hue.$2.toInt()}",
                      ),
                    ),
                    RangeSlider(
                      min: 0,
                      max: 360,
                      values: RangeValues(_hue.$1, _hue.$2),
                      onChanged: (value) {
                        setState(() {
                          _hue = (value.start, value.end);
                        });
                      },
                    ),
                    SizedBox(
                      height: 24,
                      child: Container(
                        width: double.infinity,
                        padding: const EdgeInsets.symmetric(horizontal: 20),
                        height: 24,
                        child: Container(
                          decoration: BoxDecoration(
                            borderRadius: BorderRadius.circular(4),
                            gradient: LinearGradient(
                              colors: List.generate(
                                360 ~/ 10,
                                (index) {
                                  final hue = index * 10;
                                  return HSVColor.fromAHSV(
                                    1,
                                    hue.toDouble(),
                                    _saturation,
                                    _lightness,
                                  ).toColor();
                                },
                              ),
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
            spacing,
            SliverToBoxAdapter(
              child: MaxWidth(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.start,
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 20),
                      child: Text(
                        "Saturation: ${_saturation.toStringAsFixed(2)}",
                      ),
                    ),
                    Slider(
                      min: 0.1,
                      max: 1,
                      value: _saturation,
                      onChanged: (value) {
                        setState(() {
                          _saturation = value;
                        });
                      },
                    ),
                    SizedBox(
                      height: 24,
                      child: Container(
                        width: double.infinity,
                        padding: const EdgeInsets.symmetric(horizontal: 20),
                        height: 24,
                        child: Container(
                          decoration: BoxDecoration(
                            borderRadius: BorderRadius.circular(4),
                            gradient: LinearGradient(
                              colors: List.generate(
                                360 ~/ 10,
                                (index) {
                                  return HSVColor.fromAHSV(
                                    1,
                                    hsv.hue,
                                    _saturation - 0.001,
                                    _lightness - 0.001,
                                  ).toColor();
                                },
                              ),
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
            spacing,
            SliverToBoxAdapter(
              child: MaxWidth(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.start,
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 20),
                      child: Text(
                        "Lightness: ${_lightness.toStringAsFixed(2)}",
                      ),
                    ),
                    Slider(
                      min: 0.1,
                      max: 0.9,
                      value: _lightness,
                      onChanged: (value) {
                        setState(() {
                          _lightness = value;
                        });
                      },
                    ),
                    SizedBox(
                      height: 24,
                      child: Container(
                        width: double.infinity,
                        padding: const EdgeInsets.symmetric(horizontal: 20),
                        height: 24,
                        child: Container(
                          decoration: BoxDecoration(
                            borderRadius: BorderRadius.circular(4),
                            gradient: const LinearGradient(
                              colors: [
                                Colors.black,
                                Colors.white,
                              ],
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
            SliverToBoxAdapter(
              child: MaxWidth(
                child: Padding(
                  padding: const EdgeInsets.all(20),
                  child: Column(children: [
                    Text(
                      "Color: ${hashed.toHexString()}",
                      style: TextStyle(
                        color: hashed.toColor(),
                        fontSize: 36,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    const SizedBox(
                      height: 24,
                    ),
                    Container(
                      height: 256,
                      width: double.infinity,
                      decoration: BoxDecoration(
                          borderRadius:
                              const BorderRadius.all(Radius.circular(8)),
                          color: hashed.toColor(),
                          boxShadow: [
                            BoxShadow(
                              color: hashed.toColor().withOpacity(1),
                              blurRadius: 64,
                              spreadRadius: -24,
                              offset: const Offset(0, 24),
                            )
                          ]),
                    )
                  ]),
                ),
              ),
            )
          ],
        ),
        bottomNavigationBar: Container(
          height: 64,
          color: hashed.toColor(),
          child: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Text(
                  "Copyright © ${DateTime.now().year} - Alek Angelov",
                  style: TextStyle(
                    color: hashed.toColor().computeLuminance() > 0.5
                        ? Colors.black
                        : Colors.white,
                    fontSize: 14,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                Link(
                  uri: Uri.parse("https://alekangelov.com"),
                  builder: (c, f) => TextButton(
                    onPressed: f,
                    style: TextButton.styleFrom(
                      padding: EdgeInsets.zero,
                    ),
                    child: Text(
                      "https://alekangelov.com",
                      style: TextStyle(
                        color: hashed.toColor().computeLuminance() > 0.5
                            ? Colors.black
                            : Colors.white,
                        fontSize: 14,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}


flutter_service_worker.js
'use strict';
const MANIFEST = 'flutter-app-manifest';
const TEMP = 'flutter-temp-cache';
const CACHE_NAME = 'flutter-app-cache';

const RESOURCES = {"assets/AssetManifest.bin": "693635b5258fe5f1cda720cf224f158c",
"assets/AssetManifest.json": "2efbb41d7877d10aac9d091f58ccd7b9",
"assets/FontManifest.json": "dc3d03800ccca4601324923c0b1d6d57",
"assets/fonts/MaterialIcons-Regular.otf": "2e611e8b85d96c9276d5ada334c816c1",
"assets/NOTICES": "7c0e83f91cf6a85f3b2e143899b0f784",
"assets/packages/cupertino_icons/assets/CupertinoIcons.ttf": "57d849d738900cfd590e9adc7e208250",
"assets/shaders/ink_sparkle.frag": "f8b80e740d33eb157090be4e995febdf",
"canvaskit/canvaskit.js": "76f7d822f42397160c5dfc69cbc9b2de",
"canvaskit/canvaskit.wasm": "f48eaf57cada79163ec6dec7929486ea",
"canvaskit/chromium/canvaskit.js": "8c8392ce4a4364cbb240aa09b5652e05",
"canvaskit/chromium/canvaskit.wasm": "fc18c3010856029414b70cae1afc5cd9",
"canvaskit/skwasm.js": "1df4d741f441fa1a4d10530ced463ef8",
"canvaskit/skwasm.wasm": "6711032e17bf49924b2b001cef0d3ea3",
"canvaskit/skwasm.worker.js": "19659053a277272607529ef87acf9d8a",
"favicon.png": "5dcef449791fa27946b3d35ad8803796",
"flutter.js": "6b515e434cea20006b3ef1726d2c8894",
"icons/Icon-192.png": "ac9a721a12bbc803b44f645561ecb1e1",
"icons/Icon-512.png": "96e752610906ba2a93c65f8abe1645f1",
"icons/Icon-maskable-192.png": "c457ef57daa1d16f64b27b786ec2ea3c",
"icons/Icon-maskable-512.png": "301a7604d45b3e739efc881eb04896ea",
"index.html": "373fe5cb67da4e1a3b9bb19f771252e5",
"/": "373fe5cb67da4e1a3b9bb19f771252e5",
"logo.jpg": "1d1079e01578435a05fb5578f451e500",
"main.dart.js": "3d926b2811bda10c940617a6240885f6",
"manifest.json": "9a0de1ff26661cd281cb9deb3db33b02",
"version.json": "ff966ab969ba381b900e61629bfb9789"};
// The application shell files that are downloaded before a service worker can
// start.
const CORE = ["main.dart.js",
"index.html",
"assets/AssetManifest.json",
"assets/FontManifest.json"];

// During install, the TEMP cache is populated with the application shell files.
self.addEventListener("install", (event) => {
  self.skipWaiting();
  return event.waitUntil(
    caches.open(TEMP).then((cache) => {
      return cache.addAll(
        CORE.map((value) => new Request(value, {'cache': 'reload'})));
    })
  );
});
// During activate, the cache is populated with the temp files downloaded in
// install. If this service worker is upgrading from one with a saved
// MANIFEST, then use this to retain unchanged resource files.
self.addEventListener("activate", function(event) {
  return event.waitUntil(async function() {
    try {
      var contentCache = await caches.open(CACHE_NAME);
      var tempCache = await caches.open(TEMP);
      var manifestCache = await caches.open(MANIFEST);
      var manifest = await manifestCache.match('manifest');
      // When there is no prior manifest, clear the entire cache.
      if (!manifest) {
        await caches.delete(CACHE_NAME);
        contentCache = await caches.open(CACHE_NAME);
        for (var request of await tempCache.keys()) {
          var response = await tempCache.match(request);
          await contentCache.put(request, response);
        }
        await caches.delete(TEMP);
        // Save the manifest to make future upgrades efficient.
        await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
        // Claim client to enable caching on first launch
        self.clients.claim();
        return;
      }
      var oldManifest = await manifest.json();
      var origin = self.location.origin;
      for (var request of await contentCache.keys()) {
        var key = request.url.substring(origin.length + 1);
        if (key == "") {
          key = "/";
        }
        // If a resource from the old manifest is not in the new cache, or if
        // the MD5 sum has changed, delete it. Otherwise the resource is left
        // in the cache and can be reused by the new service worker.
        if (!RESOURCES[key] || RESOURCES[key] != oldManifest[key]) {
          await contentCache.delete(request);
        }
      }
      // Populate the cache with the app shell TEMP files, potentially overwriting
      // cache files preserved above.
      for (var request of await tempCache.keys()) {
        var response = await tempCache.match(request);
        await contentCache.put(request, response);
      }
      await caches.delete(TEMP);
      // Save the manifest to make future upgrades efficient.
      await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
      // Claim client to enable caching on first launch
      self.clients.claim();
      return;
    } catch (err) {
      // On an unhandled exception the state of the cache cannot be guaranteed.
      console.error('Failed to upgrade service worker: ' + err);
      await caches.delete(CACHE_NAME);
      await caches.delete(TEMP);
      await caches.delete(MANIFEST);
    }
  }());
});
// The fetch handler redirects requests for RESOURCE files to the service
// worker cache.
self.addEventListener("fetch", (event) => {
  if (event.request.method !== 'GET') {
    return;
  }
  var origin = self.location.origin;
  var key = event.request.url.substring(origin.length + 1);
  // Redirect URLs to the index.html
  if (key.indexOf('?v=') != -1) {
    key = key.split('?v=')[0];
  }
  if (event.request.url == origin || event.request.url.startsWith(origin + '/#') || key == '') {
    key = '/';
  }
  // If the URL is not the RESOURCE list then return to signal that the
  // browser should take over.
  if (!RESOURCES[key]) {
    return;
  }
  // If the URL is the index.html, perform an online-first request.
  if (key == '/') {
    return onlineFirst(event);
  }
  event.respondWith(caches.open(CACHE_NAME)
    .then((cache) =>  {
      return cache.match(event.request).then((response) => {
        // Either respond with the cached resource, or perform a fetch and
        // lazily populate the cache only if the resource was successfully fetched.
        return response || fetch(event.request).then((response) => {
          if (response && Boolean(response.ok)) {
            cache.put(event.request, response.clone());
          }
          return response;
        });
      })
    })
  );
});
self.addEventListener('message', (event) => {
  // SkipWaiting can be used to immediately activate a waiting service worker.
  // This will also require a page refresh triggered by the main worker.
  if (event.data === 'skipWaiting') {
    self.skipWaiting();
    return;
  }
  if (event.data === 'downloadOffline') {
    downloadOffline();
    return;
  }
});
// Download offline will check the RESOURCES for all files not in the cache
// and populate them.
async function downloadOffline() {
  var resources = [];
  var contentCache = await caches.open(CACHE_NAME);
  var currentContent = {};
  for (var request of await contentCache.keys()) {
    var key = request.url.substring(origin.length + 1);
    if (key == "") {
      key = "/";
    }
    currentContent[key] = true;
  }
  for (var resourceKey of Object.keys(RESOURCES)) {
    if (!currentContent[resourceKey]) {
      resources.push(resourceKey);
    }
  }
  return contentCache.addAll(resources);
}
// Attempt to download the resource online before falling back to
// the offline cache.
function onlineFirst(event) {
  return event.respondWith(
    fetch(event.request).then((response) => {
      return caches.open(CACHE_NAME).then((cache) => {
        cache.put(event.request, response.clone());
        return response;
      });
    }).catch((error) => {
      return caches.open(CACHE_NAME).then((cache) => {
        return cache.match(event.request).then((response) => {
          if (response != null) {
            return response;
          }
          throw error;
        });
      });
    })
  );
}


lib/color_hash.dart
// string to sha256 to color
import 'dart:ui';

_hueToRgb(
  double p,
  double q,
  double t,
) {
  if (t < 0) {
    t += 1;
  }
  if (t > 1) {
    t -= 1;
  }
  if (t < 1 / 6) {
    return p + (q - p) * 6 * t;
  }
  if (t < 1 / 2) {
    return q;
  }
  if (t < 2 / 3) {
    return p + (q - p) * (2 / 3 - t) * 6;
  }
  return p;
}

// hsl to rgb

_hslToRgb(
  double h,
  double s,
  double l,
) {
  double r, g, b;

  // there's no saturation, it's an achromatic color
  if (s == 0) {
    r = g = b = l;
    return (r, g, b);
  }

  final q = l < 0.5 ? l * (1 + s) : l + s - l * s;
  final p = 2 * l - q;
  r = _hueToRgb(p, q, h + 1 / 3);
  g = _hueToRgb(p, q, h);
  b = _hueToRgb(p, q, h - 1 / 3);

  return (r, g, b);
}

/// A tuple of two numbers.
typedef MinMax = (double, double);

/// A function that takes a string and returns a [Color].
///
/// This is used to create a [ColorHash] with custom settings.
///
/// See [ColorHash.createHasher].
typedef Hasher = Color Function(String);

class ColorHash {
  /// Creates a [ColorHash] from a string.
  ///
  /// [lightness] and [saturation] are used to control the saturation and lightness of the generated color.
  ///
  /// [hue] is used to control the hue of the generated color.
  ///
  /// [hue] must be a tuple of two numbers between 0 and 360.
  ///
  /// throws [AssertionError] if any of the values are invalid.
  ///
  /// throws [AssertionError] if [hue] is not a tuple of two numbers.
  ///
  /// throws [AssertionError] if [hue] values are not between 0 and 360.
  ///
  /// throws [AssertionError] if [lightness] or [saturation] are not between 0 and 1.
  ///
  /// ```dart
  /// final colorHash = ColorHash("Some string");
  /// Color c = colorHash.toColor()
  /// ```
  ColorHash(
    this.str, {
    this.lightness = 0.5,
    this.saturation = 0.5,
    this.hue = (0, 360),
  })  : assert(lightness >= 0 && lightness <= 1),
        assert(saturation >= 0 && saturation <= 1),
        assert(hue.$1 >= 0 && hue.$1 <= 360),
        assert(hue.$2 >= 0 && hue.$2 <= 360),
        assert(hue.$1 < hue.$2);

  final String str;
  final double lightness;
  final double saturation;
  final MinMax hue;
  late final Color color = toColor();

  /// Returns a [Color] from the string.
  ///
  /// returns a [Color] from the string.
  ///
  /// ```dart
  /// final colorHash = ColorHash("Some string");
  /// Color c = colorHash.toColor()
  /// ```
  Color toColor() {
    var H = str.hashCode;
    var S = saturation;
    var L = lightness;
    final (min, max) = hue;
    var h = min + (H % (max - min));
    var (r, g, b) = _hslToRgb(h / 360, S, L);
    r = (r * 255).round();
    g = (g * 255).round();
    b = (b * 255).round();
    final c = Color.fromRGBO(r, g, b, 1);
    return c;
  }

  @override
  String toString() {
    return 'ColorHash{color: $color}';
  }

  String toHexString() {
    final red = color.red.toRadixString(16).toUpperCase();
    final green = color.green.toRadixString(16).toUpperCase();
    final blue = color.blue.toRadixString(16).toUpperCase();
    final alpha = color.alpha.toRadixString(16).toUpperCase();
    return '#$red$green$blue${alpha == "FF" ? "" : alpha}';
  }

  /// Creates a [Hasher] with custom settings.
  ///
  /// [lightness] and [saturation] are used to control the saturation and lightness of the generated color.
  ///
  /// [hue] is used to control the hue of the generated color.
  ///
  /// [hue] must be a tuple of two numbers between 0 and 360.
  ///
  /// throws [ArgumentError] if any of the values are invalid.
  ///
  /// throws [ArgumentError] if [hue] is not a tuple of two numbers.
  ///
  /// throws [ArgumentError] if [hue] values are not between 0 and 360.
  ///
  /// throws [ArgumentError] if [lightness] or [saturation] are not between 0 and 1.
  ///
  /// returns a [Hasher] function that takes a string and returns a [Color].
  ///
  /// ```dart
  /// final hasher = ColorHash.createHasher(
  ///     lightness: 0.9,
  ///     saturation: 0.25,
  ///     hue: (20, 30),
  /// );
  /// final color = hasher("Some string");
  /// ```
  static Hasher createHasher({
    double lightness = 0.5,
    double saturation = 0.5,
    MinMax hue = (0, 360),
  }) {
    if (lightness < 0 || lightness > 1) {
      throw ArgumentError.value(
        lightness,
        'lightness',
        'lightness must be between 0 and 1',
      );
    }
    if (saturation < 0 || saturation > 1) {
      throw ArgumentError.value(
        saturation,
        'saturation',
        'saturation must be between 0 and 1',
      );
    }
    if (hue.$1 < 0 || hue.$1 > 360 || hue.$2 < 0 || hue.$2 > 360) {
      throw ArgumentError.value(
        hue,
        'hue',
        'hue values must be between 0 and 360',
      );
    }
    return (String str) => ColorHash(
          str,
          lightness: lightness,
          saturation: saturation,
          hue: hue,
        ).toColor();
  }
}
